core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/NullOrEmptyCollectionAdapter.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.dto.server;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+import java.lang.reflect.Type;
+import java.util.List;
+
+/**
+ * This class will prevent serialization of null or empty fields of DTOs with Collection types
+ *
+ * @author Mykhailo Kuznietsov
+ */
+public class NullOrEmptyCollectionAdapter implements JsonSerializer<List<?>> {
+  @Override
+  public JsonElement serialize(List<?> src, Type typeOfSrc, JsonSerializationContext context) {
+    if (src == null || src.isEmpty()) {
+      return null;
```

Comment:
I'm not sure this is correct. Isn't there a semantic difference between `null` and empty list? While it may solve an issue for stuff within the devfile, I'm not sure it won't cause trouble for other types of entities...

Reply:
According to our DTOs impl - there is not any semantic difference between null or empty list/map. I mean null and empty collection/map can be used only as initialized collection/map (not null) after parsing.
From Che Server and GWT (client impls) perspective, nothing is changed.
For CLI that we used to use nothing is changed as well since typescript DTO generator initializes null fields.

The things that may be affected: 
- UD: it uses received JSON directly via interface without using DTO impls, and null collections/maps are not initialized. So, code like `workspace.devfile.properties["persistVolumes"] = false` may cause a failure, but I'm not sure. Thank @ashumilova, who agreed to check this and provide more info how it may affect UD;
- Any other Che Server API client that relies on current DTO serialization style where all fields are set, even if they were not set initially;

I think we can stop serialization empty collections/maps if does not cause any ingressions.
What are the alternatives?:
- stop using the same DTO object for all kind of components types and have like different classes for CheEditorComponentDto, ChePluginComponentDto, .... I do not have a clear picture of how it could be implemented;
- make ComponentDto - a special DTO that has untypical behavior for DTOs (null, empty collections/maps are not serialized). It looks safer against the implemented solution, but it makes our DTOs behavior variational from class to class, which is not a good thing IMHO.
- Do not use DTOs for devfile at all, and it's not clear what we can use instead, especially if we take into account that Devfile should be returned as part of WorkspaceDto.

> I'm not sure it won't cause trouble for other types of entities...

@metlos Can you elaborate more about your thoughts about possible issues? Are you OK with this approach if everything works fine (ci-test result is 97+%, I do not expect issues in master)?

Reply:
Well, the thing that changes for the clients is that where they saw empty collections coming from our API endpoints, they're going to start seeing nulls. So if anyone happens to rely on non-null arrays being present, they will be caught by surprise by this...

Reply:
in case of objects, instead of properties with empty arrays, they won't see the properties at all anymore, so the clients will have to start dealing with undefined values where before they didn't have to. Am I right?

Reply:
I'm OK with making such change because we kinda need it, I just want to point out the consequences for the existing clients (e.g. che-theia and any other custom UI built on top of wsmaster API).

Reply:
> so the clients will have to start dealing with undefined values where before they didn't have to. Am I right?

Right!

Reply:
@mkuznyetsov and I found some generated typescript DTOs impl which look like initialize collections and maps
```
home/michael/IdeaProjects/che/dockerfiles/lib/src/api/dto/che-dto.ts

export module org.eclipse.che.api.workspace.shared.dto.devfile {

 export class DevfileDtoImpl implements org.eclipse.che.api.workspace.shared.dto.devfile.DevfileDto {

       components : Array<org.eclipse.che.api.workspace.shared.dto.devfile.ComponentDto>;
       metadata : org.eclipse.che.api.workspace.shared.dto.devfile.MetadataDto;
       projects : Array<org.eclipse.che.api.workspace.shared.dto.devfile.ProjectDto>;
       apiVersion : string;
       name : string;
       attributes : Map<string,string>;
       commands : Array<org.eclipse.che.api.workspace.shared.dto.devfile.DevfileCommandDto>;

   __jsonObject : any;

   constructor(__jsonObject?: any) {
     this.__jsonObject = __jsonObject;
           this.components = new Array<org.eclipse.che.api.workspace.shared.dto.devfile.ComponentDto>();
           if (__jsonObject) {
             if (__jsonObject.components) {
                 __jsonObject.components.forEach((item) => {
                 this.components.push(new org.eclipse.che.api.workspace.shared.dto.devfile.ComponentDtoImpl(item));
                 });
             }
           }
```
@mmorhun helps us to find that Theia works with DTOs without impls but via generated interfaces, but undefined/null should not happen here since all fields are marked as optional (`?`) and `strictNull` option is set on compiler, everyplace  should contain the corresponding check before invocation of it. Terminology may be not right, but that's what I understand from converstation with @evidolob and @mmorhun 

And Dashboard uses manually written interfaces and not all fields are marked as optional
https://github.com/eclipse/che/blob/master/dashboard/src/components/typings/che.d.ts#L340
Also, I do not know what is settings of compiler, if before invocation client should check field existence. @ashumilova @benoitf Can you take a look? For Dashboard it works in the same way as for Theia? If not, can we so them work in the same way?

Reply:
in dashboard it fails at runtime if try to access `workspace.attributes.plugins`, when attributes is null.

Reply:
For typescript, consumers of the DTO should never see null values on collections/map

if field is mandatory then it needs to be `empty` value, if the field is optional then it can be `undefined`

if it's not the case, then the TypeScript DTO needs to be enhanced.

in Ann's case, if attributes is optional field then all consumers have to check if it's defined or not but if attributes is mandatory then it should always return an empty value